HHS/Head-Start-TTADP
# Add GTM snippet

## Description of change

Adds Google Tag Manager snippet with Google analytics measurement. Biggest change to the site is the frontend `index.html` has a `nonce` replaced when retrieved from the server.

From https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_inline_script we must use `unsafe-inline`, provide a `nonce`, or provide the hash of the script for that script to execute. The point of google tag manager is the contents of the scripts can change without a software deployment (via CircleCI) so we can't generate a hash at build time (because the hash could be different when users are retrieving the page). Also see https://developers.google.com/tag-platform/tag-manager/web/csp#enable_the_container_tag_to_use_csp.

Now the frontend `index.html` has a `__NONCE__` in the tag manager script snippet. When `index.html` is retrieved from the server `__NONCE__` is replaced with a generated `nonce` value. The same `nonce` is placed in the response `csp` header.

Google tag manager and google analytics were setup with environments that match ours, local, sandbox, dev, staging, and prod (or GA calls it live). There are build variables in CircleCI that insert the correct tag manager options into the `index.html` page depending on the environment. There is a "lookup" variable in google tag manager that matches the google tag manager environment to the proper google analytics `Measurement ID`.

NOTE: Google analytics wouldn't let me name a property with three characters so `dev` is `development`

## How to test

This is currently deployed to sandbox. When you browse to sandbox you should see the google tag manager script in your network tab and after browsing around calls to google analytics. Login to https://analytics.google.com/ and https://tagmanager.google.com/ using the google account in our `1password` vault. You should see your session inside google analytics when viewing the sandbox environment. Some more things to check, sorry they aren't easy to navigate to, let me know if you need help finding:

 * Environment variables in circle ci only show the last four characters but double check those characters match those in google tag manager. Find the values in tag manager by clicking the `Admin` tab, environments, `Actions` context menu, and `Get Snippet`. You'll have to search a bit but they are the query params `gtm_auth` and `gtm_preview`
 * Make sure the mapping between google tag manager and google analytics is correct. Can check by clicking `Variables` in the left hand menu, and selecting `Lookup - GA Tracking IDs`. Can find the values in google analytics by clicking the gear icon in the lower left, `Data Streams`, click the only option, the `Use existing on-page tag`, `Google Analytics`, and the `Measurement ID` can be found on the fourth step

Once dev is no longer in use I'll deploy this branch to both to ensure we can get analytics for `dev` and `sandbox` separately.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-379
* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-791

## Checklists

### Every PR


- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [n/a] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy


### After merge/deploy
